### PR TITLE
Settings watcher false alarms on boot [ESD-1049]

### PR DESCRIPTION
**Pull request by pmiettinen**
Requires https://github.com/swift-nav/piksi_buildroot/pull/1163

### CODE REVIEW

File: src/setting_sbp_cb.c [hunk 1]
```diff
@@ -25,25 +25,45 @@
 #include <internal/setting_def.h>
 #include <internal/setting_sbp_cb.h>
 
+#define UPDATE_FILTER_NONE 0x0
+#define UPDATE_FILTER_READONLY 0x1
+#define UPDATE_FILTER_WATCHONLY 0x2
```
Comment (benjaminaltieri) [inline src/setting_sbp_cb.c]: possibly clearer with bit shifting to indicate usage as mask/flags
```suggestion
#define UPDATE_FILTER_READONLY (0x1 << 1)
#define UPDATE_FILTER_WATCHONLY (0x1 << 2)

#define update_filter_check(filter_mask, filter_enum)  (0 == (filter_mask & filter_enum))
```
encapsulate checking either with macro or fully fledged helper func

File: src/setting_sbp_cb.c [hunk 2]
```diff
@@ -61,28 +81,20 @@ static void setting_update_value(settings_t *ctx, uint8_t *msg, uint8_t len)
     return;
   }
 
-  if (setting_data->watchonly) {
+  if ((filter & UPDATE_FILTER_WATCHONLY) && setting_data->watchonly) {
```
Comment (benjaminaltieri) [inline src/setting_sbp_cb.c]:
```suggestion
  if (update_filter_check(filter, UPDATE_FILTER_WATCHONLY) && setting_data->watchonly) {
```
just an illustration of the above suggestion

File: src/setting_sbp_cb.c [hunk 3]
```diff
@@ -116,7 +128,14 @@ static void setting_register_resp_callback(uint16_t sender_id,
...
 
-  setting_update_value(ctx, (uint8_t *)resp->setting, len - sizeof(resp->status));
+  /* Update the actual setting. In case of readonly, trust the initialized value. */
+  setting_value_update(ctx,
+                       resp->setting,
+                       len - sizeof(resp->status),
+                       UPDATE_FILTER_WATCHONLY | UPDATE_FILTER_READONLY);
+
+  /* Update watchers, no need for filter as wathers shall not be read-only */
+  setting_value_update(ctx, resp->setting, len - sizeof(resp->status), UPDATE_FILTER_NONE);
```
Comment (benjaminaltieri) [inline src/setting_sbp_cb.c]: not understanding maybe, but why the double call? the only difference I see between the two is early return if either read only or watcher, but then the second call would follow the logic previously avoided with early returns?

File: src/setting_sbp_cb.c [hunk 4]
```diff
@@ -220,15 +213,16 @@ static void setting_write_resp_callback(uint16_t sender_id,
     return;
   }
 
-  if (setting_update_watch_only(ctx, write_response->setting, len - sizeof(write_response->status))
-      != 0) {
-    log_warn("error in settings read response message");
-  }
+  /* Update watchers, no need for filter as wathers shall not be read-only */
```
Comment (benjaminaltieri) [inline src/setting_sbp_cb.c]: wathers

File: src/setting_sbp_cb.c [hunk 5]
```diff
@@ -116,7 +128,14 @@ static void setting_register_resp_callback(uint16_t sender_id,
...
   }
 
-  setting_update_value(ctx, (uint8_t *)resp->setting, len - sizeof(resp->status));
+  /* Update the actual setting. In case of readonly, trust the initialized value. */
+  setting_value_update(ctx,
+                       resp->setting,
+                       len - sizeof(resp->status),
+                       UPDATE_FILTER_WATCHONLY | UPDATE_FILTER_READONLY);
+
+  /* Update watchers, no need for filter as wathers shall not be read-only */
```
Comment (benjaminaltieri) [inline src/setting_sbp_cb.c]: wathers

File: src/setting_sbp_cb.c [hunk 6]
```diff
@@ -220,15 +213,16 @@ static void setting_write_resp_callback(uint16_t sender_id,
...
 
-  if (setting_update_watch_only(ctx, write_response->setting, len - sizeof(write_response->status))
-      != 0) {
-    log_warn("error in settings read response message");
-  }
+  /* Update watchers, no need for filter as wathers shall not be read-only */
+  setting_value_update(ctx,
+                       write_response->setting,
+                       len - sizeof(write_response->status),
+                       UPDATE_FILTER_NONE);
```
Comment (benjaminaltieri) [inline src/setting_sbp_cb.c]: this refactor doesn't quite retain the original intent of `setting_update_watch_only` which was to write the value to a setting if and only if that setting is watch only. now that the code is centralized it is much easier to reason about this so I'm less worried. maybe we can just add a `UPDATE_FILTER_NORMAL` which I suppose would return on the following condition: `!setting_data->readonly && !setting_data->watchonly`. open to other ideas too

File: src/settings.c [hunk 7]
```diff
@@ -357,7 +358,11 @@ static int setting_read_watched_value(settings_t *ctx, setting_data_t *setting_d
                                               WATCH_INIT_TIMEOUT_MS,
                                               WATCH_INIT_TRIES,
                                               SBP_SENDER_ID,
-                                              NULL);
+                                              &req_state);
+
+  if (0 == result && strlen(req_state.resp_value) > 0) {
```
[the remark this answers is not in the document]
Comment (benjaminaltieri) [reply src/settings.c]: also some explanation that in the case `(0 == result && strlen(req_state.resp_value) == 0)` then `setting_read_watched_value` returns success even though no value was read into memory

File: src/setting_sbp_cb.c (hunk 4 above)
Comment (pmiettinen) [reply src/setting_sbp_cb.c]: 🔨

File: src/setting_sbp_cb.c (hunk 5 above)
Comment (pmiettinen) [reply src/setting_sbp_cb.c]: 🛠

File: src/settings.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (pmiettinen) [reply src/settings.c]: Also, this doesn't cover valid empty string.

File: src/setting_sbp_cb.c [hunk 8]
```diff
@@ -25,25 +25,45 @@
...
-  if (api->send(api->ctx, SBP_MSG_SETTINGS_WRITE_RESP, len, (uint8_t *)write_response) != 0) {
+  if (api->send(api->ctx, SBP_MSG_SETTINGS_WRITE_RESP, resp_len, (uint8_t *)write_response) != 0) {
     log_error("sending settings write response failed");
     return -1;
   }
   return 0;
 }
 
-static void setting_update_value(settings_t *ctx, uint8_t *msg, uint8_t len)
+static void setting_value_update(settings_t *ctx, const char *msg, uint8_t len, uint32_t filter)
```
[the remark this answers is not in the document]
Comment (pmiettinen) [reply src/setting_sbp_cb.c]: ⚒

File: src/setting_sbp_cb.c (hunk 3 above)
Comment (pmiettinen) [reply src/setting_sbp_cb.c]: True, this is a blunder based on earlier version.

File: src/setting_sbp_cb.c (hunk 6 above)
Comment (pmiettinen) [reply src/setting_sbp_cb.c]: Seems fine for now. If there's anymore properties in the future than read/watchonly, then reconsider.

**Comment by pmiettinen** [comment]
@benjaminaltieri - Your comments should be processed. Additionally I noticed a race condition if signaling is done from `request_state_check()` and the response is still unprocessed in the actual callback.

**Pull request merged**